Documentation overhaul (#75)
greenbone/eulabeia: docs/requirements.md

# Requirements

This document describes the requirements. 

> The favored usage of eulabeia is via [docker](#Docker).

Author: I am pretty sure that the local documentation will be outdated pretty soon but that we will keep the docker images running; hence it is favored. As a counter part to your statement: I just use docker images without having it installed locally at all.

It is divided between two methods:
- [docker](#Docker), will describe how to setup docker and use the provided makefile
- [local](#Local), will describe how to setup and install the requirements without provided images

This document will use Debian 11 as a basis and is targeting developer machines with relatively lax security requirements. 

> For production usage please harden docker, redis and mosquitto accordingly.

Author: I am not sure why are not sure what I mean by this. It is means: the configuration of the services within document is insecure; do not use it in production.

## Docker

To use eulabeia with docker you need to install `go`, `docker` as well as `gnu-make` to make it easier to use you can add your user to the docker group.

```
# apt-get install docker.io make golang

> # usermod -a -G docker your_username

Author: yes

> # usermod -a -G docker your_username

Reviewer: ok

```

To reload the newly assign group, login to your user account again

```
$ su - $USER
$ docker ps
```

to build the images based on your changes you can use `make start-container` with that you have a useable sensor, director and broker up and running.

If you want to verify the setup you can simply use `make run-example-client`.

If you have `mosquitto-clients` installed, you can can verify that the local connection works:
```
$ mosquitto_sub -L 'mqtt://localhost:9138/scanner/#'
```

to subscribe to any message inside the scanner context. 

Afterwards you can send a irregular message

```
$ mosquitto_pub -L 'mqtt://localhost:9138/scanner/scan/cmd/director' -m "hi"
```

### Broker

> To run a broker you can execute `make start-broker` this will run mosquitto and open a local port 9138.

Reviewer: ```suggestion
To run a broker you can execute `make start-broker`. This will run mosquitto and opens the local port `9138`.
```

### Sensor

To run a sensor you can execute `make start-sensor` this will run mosquitto and open a local port 9138 as well as start a sensor containing a test nasl file (oid: `1.3.6.1.4.1.25623.1.0.90022`) used by openvas.

> To communicate with the sensor use mqtt as described previously.

Reviewer: ```suggestion
To communicate with the sensor use MQTT as described previously.
```

### Director

> To run a director you can execute `make start-director` this will run mosquitto and open a local port 9138 as well as start a director.

Reviewer: ```suggestion
To run a director you can execute `make start-director`. This will run the mosquitto broker and opens the local port `9138` as well as start a director.
```

To communicate with the director use mqtt as described previously.

## Local

### Broker (mosquitto)

To install and configure mosquitto:

```
# apt-get install mosquitto
# echo "allow_anonymous true" > /etc/mosquitto/conf.d/allow_anonymous.conf
# systemctl restart mosquitto
```

This allows anonymous connections into the running mqtt broker.

### Redis

Redis is used by openvas as well as sensor.

To install and configure redis:

```

> # apt install redis
> # sed -i 's/\# unixsocket\([ p]\)/unixsocket\1/' /etc/redis/redis.conf
> # systemctl restart redis
> # usermod -a -G redis your_username

Reviewer: ```suggestion
sudo apt install redis
sudo sed -i 's/\# unixsocket\([ p]\)/unixsocket\1/' /etc/redis/redis.conf
sudo systemctl restart redis
sudp usermod -a -G redis $USER
```

```

This should create a redis socket in `/run/redis/redis-server.sock`

### gvm-libs

> Is used by libeulabeia to establish mqtt-connection and openvas.
> 
> ```
> # apt install libglib2.0-dev \

Reviewer: ```suggestion
Is used by `libeulabeia` to establish mqtt-connection and [openvas](https://github.com/greenbone/openvas-scanner).

```
sudo apt install libglib2.0-dev \
```

    libical-dev \
    libnet1-dev \
    libnet1-dev \
    libpaho-mqtt-dev \
    libpcap-dev \
    libpq-dev \
    libssh-gcrypt-dev \
    libssl-dev \
    libxml2-dev \
    make \
    pkg-config \
    postgresql-server-dev-all \
    uuid-dev \
    xsltproc
```

```

$ git clone https://github.com/greenbone/gvm-libs.git
$ git checkout -t origin/middleware
$ cmake -Bbuild -DREDIS_SOCKET_PATH=/run/redis/redis-server.sock
$ sudo 'cmake --build build --target install'
$ sudo 'ldconfig'
```

### libeulabeia

Is used by openvas to communicate with eulabeia.

```
# apt install libjson-glib-dev
```

```
$ make c-library
$ sudo 'cmake --build c/libeulabeia/build -- install'
$ sudo 'ldconfig'
```

### openvas

openvas is used by our sensor.

```
# apt install libksba-dev bison
```

```
$ git clone https://github.com/greenbone/openvas-scanner.git
$ cmake -Bbuild
$ sudo 'cmake --build build --target install'
$ echo "db_address = /run/redis/redis-server.sock" | sudo tee /etc/openvas/openvas.conf
$ echo "mqtt_server_uri = localhost:1883" | sudo tee -a /etc/openvas/openvas.conf
$ echo "$USER ALL = NOPASSWD: /usr/local/sbin/openvas" | sudo tee /etc/sudoers.d/allow_openvas
```

### director

```

> # apt install golang

Reviewer: ```suggestion
sudo apt install golang
```

```

```

> $ make build-director
> $ mkdir -p ~/.config/eulabeia
> $ sed 's/server = ".*/server = "localhost:1883"/' config.toml | tee ~/.config/eulabeia/config.toml
> $ sed -i 's/StoragePath = ".*/StoragePath = "\/tmp\/"/' ~/.config/eulabeia/config.toml
> $ sed -i 's/KeyFile = ".*/KeyFile = "\/tmp\/private.key"/' ~/.config/eulabeia/config.toml
> $ ./bin/eulabeia-director

Reviewer: ```suggestion
make build-director
mkdir -p ~/.config/eulabeia
sed 's/server = ".*/server = "localhost:1883"/' config.toml | tee ~/.config/eulabeia/config.toml
sed -i 's/StoragePath = ".*/StoragePath = "\/tmp\/"/' ~/.config/eulabeia/config.toml
sed -i 's/KeyFile = ".*/KeyFile = "\/tmp\/private.key"/' ~/.config/eulabeia/config.toml
./bin/eulabeia-director
```

```

> This changes the mqtt server address to `localhost:1883`, the storage and key file path to `/tmp`.

Reviewer: ```suggestion
This changes the MQTT server address to `localhost:1883`, as well as the storage and key file paths to `/tmp`.
```

> ### sensor
> 

Reviewer: ```suggestion
### sensor

Installing the required packages with `apt` and configure the sensor as follows:

```

```

> # apt install golang

Reviewer: ```suggestion
sudo apt install golang
```

```
sudo apt install libjson-glib-dev
```

```
make c-library
sudo cmake --build c/libeulabeia/build -- install
sudo ldconfig
```

### openvas

[openvas](https://github.com/greenbone/openvas-scanner) is used by our sensor.

```
sudo apt install libksba-dev bison
```

```
git clone https://github.com/greenbone/openvas-scanner.git -b middleware
cmake -Bbuild
sudo cmake --build build --target install
echo "db_address = /run/redis/redis-server.sock" | sudo tee /etc/openvas/openvas.conf
echo "mqtt_server_uri = localhost:1883" | sudo tee -a /etc/openvas/openvas.conf
echo "$USER ALL = NOPASSWD: /usr/local/sbin/openvas" | sudo tee /etc/sudoers.d/allow_openvas
```

### director

Installing the required packages with `apt` and configure the director as follows:

```
sudo apt install golang
```

```
make build-director
mkdir -p ~/.config/eulabeia
sed 's/server = ".*/server = "localhost:1883"/' config.toml | tee ~/.config/eulabeia/config.toml
sed -i 's/StoragePath = ".*/StoragePath = "\/tmp\/"/' ~/.config/eulabeia/config.toml

> sed -i 's/KeyFile = ".*/KeyFile = "\/tmp\/private.key"/' ~/.config/eulabeia/config.toml

Author: gvmd will use work with temporary scans so there it is not an issue; if that's the only reason I will change it to something else; any suggestions?

> sed -i 's/KeyFile = ".*/KeyFile = "\/tmp\/private.key"/' ~/.config/eulabeia/config.toml

Author: changed to `~/.config/eulabeia` for `private.key` and local
